### PR TITLE
Stamp build-In Method Support for Anydata 

### PR DESCRIPTION
**Pull request by mohanvive**
The syntax would be as below. 
```ballerina
Tx|error varName = Tx.stamp(y)
```
Inbuilt function, stamp can only be called on a type descriptor. 

Here, y would be aa anydata type argument. This also can return error based on the runtime exception that causes when stamping the value type to provided storage type. 

## Automation tests
 - Unit tests 
   > Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```ballerina
type Teacher record {
    string name;
    int age;
    string status;
    string batch;
    string school;
};

type NonAcademicStaff record {
    string name;
    int age;
    string status;
    string batch;
    string...
};

function stampOpenRecordToTypeClosedRecord() returns NonAcademicStaff|error  {
    Teacher teacher = { name: "Raja", age: 25, status: "single", batch: "LK2014", school: "Hindu College" };
    NonAcademicStaff|error  returnValue = NonAcademicStaff.stamp(teacher);

    return returnValue;
}
```

### CODE REVIEW

**Comment by mohanvive** [comment]
Code Refactoring WIP.

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 1]
```diff
@@ -729,7 +730,11 @@ expression
     |   trapExpr                                                            # trapExpression
     |	expression matchExpression										    # matchExprExpression
     |   expression ELVIS expression                                         # elvisExpression
-    |   typeName                                                            # typeAccessExpression
+    |   typeDescExpr                                                        # typeAccessExpression
+    ;
+
+typeDescExpr
+    : typeName                                                              # typeDescExpression
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: Remove the label.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 2]
```diff
@@ -915,4 +1006,67 @@ private boolean isMemberAllowed(SymbolEnv env, BSymbol symbol) {
...
+     * @return eligibility to use as the target type for 'stamp' function
+     */
+    private boolean isStampSupportedForSourceType(BType sourceType) {
+
+        if (types.isAnydata(sourceType)) {
+            switch (sourceType.tag) {
+                case TypeTags.ARRAY:
+                    // Primitive type array does not support stamp because primitive arrays are not using ref registry.
+                    int arrayConstraintTypeTag = ((BArrayType) sourceType).eType.tag;
+                    return !(arrayConstraintTypeTag == TypeTags.INT || arrayConstraintTypeTag == TypeTags.BOOLEAN ||
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: We could check whether tag is less than some value to detect these all.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -240,6 +240,150 @@ public boolean isAssignable(BType source, BType target) {
...
+                    if (((BJSONType) source).getConstraint().tag != TypeTags.NONE) {
+                        return isStampingAllowed(((BJSONType) source).getConstraint(),
+                                ((BJSONType) target).getConstraint());
+                    }
+                } else if (source.tag == TypeTags.MAP) {
+                    return isStampingAllowed(((BMapType) source).getConstraint(),
+                            ((BJSONType) target).getConstraint());
+                }
+            } else {
+                if (source.tag == TypeTags.JSON || source.tag == TypeTags.RECORD || source.tag == TypeTags.MAP) {
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Refactor this.

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 (hunk 1 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 4]
```diff
@@ -313,6 +316,94 @@ public BSymbol resolveBuiltinOperator(Name name, BType... args) {
         return bSymbol;
     }
 
+    BSymbol createSymbolForStampOperator(DiagnosticPos pos, Name name, List<BLangExpression> functionArgList,
+                                         BLangExpression targetTypeExpression) {
+        // If there are more than one argument for stamp in-built function then fail.
+        if (functionArgList.size() != 1) {
+            dlog.error(pos, DiagnosticCode.TOO_MANY_ARGS_FUNC_CALL, name);
+            resultType = symTable.semanticError;
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java (hunk 2 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 3 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal [hunk 5]
```diff
@@ -0,0 +1,93 @@
+type Employee record {
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Shall we add the license header to all the newly added ballerina test files?

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+type Teacher record {
+    string name;
+    int age;
+    string status;
+    string batch;
+    string school;
+};
+
+function stampAnydataToJSON() returns json|error  {
+
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Shall we remove this extra line?

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    return mapValue;
+}
+
+function stampAnydataToAnydata() returns anydata {
+    json jsonValue = { name: "Raja", age: 25, status: "single", batch: "LK2014", school: "Hindu College" };
+    anydata anydataValue = jsonValue;
+    anydata returnValue = anydata.stamp(anydataValue);
+
+    return returnValue;
+}
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Shall we add a new line to the end of the file?

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+function stampAnydataToJSONV2() returns json|error  {
+    json t1 = { name: "Raja", age: 25, status: "single", batch: "LK2014", school: "Hindu College" };
+    anydata anydataValue = t1;
+
+    json|error  jsonValue = json.stamp(anydataValue);
+    return jsonValue;
+}
+
+function stampAnydataToXML() returns xml|error  {
+
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Shall we remove the extra line added after each function signature?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java [hunk 9]
```diff
@@ -327,6 +330,32 @@ public BType getType() {
         return this.type;
     }
 
+    @Override
+    public void stamp(BType type) {
+
```
Comment (NatashaWso2) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java]: Shall we remove this extra line added after the function signature?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueType.java [hunk 10]
```diff
@@ -18,6 +18,8 @@
 
 package org.ballerinalang.model.values;
 
+import org.ballerinalang.model.types.BType;
+
```
Comment (NatashaWso2) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueType.java]: Shall we remove this extra line added after the import?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 11]
```diff
@@ -313,6 +316,98 @@ public BSymbol resolveBuiltinOperator(Name name, BType... args) {
...
+        return targetType;
+    }
+
+    private BSymbol resolveTargetSymbolForStamping(BType targetType, BType variableSourceType, Name name,
+                                                   DiagnosticPos pos) {
+        if (!isStampSupportedForTargetType(targetType)) {
+            dlog.error(pos, DiagnosticCode.INCOMPATIBLE_STAMP_TYPE, variableSourceType, targetType);
+            resultType = symTable.semanticError;
+            return symTable.notFoundSymbol;
+
```
Comment (NatashaWso2) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Shall we remove this extra line added after the return?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 12]
```diff
@@ -915,4 +1010,69 @@ private boolean isMemberAllowed(SymbolEnv env, BSymbol symbol) {
...
+    }
+
+    /**
+     * Returns the eligibility whether stamp can be on the given value type.
+     *
+     * @param sourceType source type used for the stamp operation
+     * @return eligibility to use as the target type for 'stamp' function
+     */
+    private boolean isStampSupportedForSourceType(BType sourceType) {
+
```
Comment (NatashaWso2) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Shall we remove the extra line added after the function signature?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java (hunk 2 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Will be incorporate with next iteration.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal (hunk 5 above)
Comment (mohanvive) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Added.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal (hunk 6 above)
Comment (mohanvive) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Fixed.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal (hunk 8 above)
Comment (mohanvive) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Fixed.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal (hunk 7 above)
Comment (mohanvive) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/stamp/anydata-stamp-expr-test.bal]: Fixed.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java (hunk 9 above)
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java]: Fixed.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueType.java (hunk 10 above)
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueType.java]: Done.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java (hunk 11 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java (hunk 12 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Fixed.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/util/codegen/Mnemonics.java [hunk 13]
```diff
@@ -79,6 +79,7 @@
         mnemonics[InstructionCodes.FREEZE] = "freeze";
         mnemonics[InstructionCodes.IS_FROZEN] = "isFrozen";
 
+        mnemonics[InstructionCodes.STAMP] = "stamp";
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/util/codegen/Mnemonics.java]: This needs to be moved before `mnemonics[InstructionCodes.FREEZE] = "freeze";` right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java [hunk 14]
```diff
@@ -599,4 +601,25 @@ private void defineOperator(Name name,
         BOperatorSymbol symbol = new BOperatorSymbol(name, rootPkgSymbol.pkgID, opType, rootPkgSymbol, opcode);
         rootScope.define(name, symbol);
     }
+
+    public BOperatorSymbol createOperator(Name name,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java]: Shouldn't this method be in the SymbolResolver?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 15]
```diff
@@ -762,6 +762,13 @@ error.binary.too.small=\
 error.partially.initialized.variable=\
   variable ''{0}'' may not have been initialized
 
+# stamp inbuilt method related error
+error.incompatible.stamp.type=\
+  Incompatible stamp type: type ''{0}'' cannot be stamped as type ''{1}''
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: Should Incompatible start with a lower case `i` instead?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/util/codegen/Mnemonics.java (hunk 13 above)
Comment (MaryamZi) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/util/codegen/Mnemonics.java]: Same with the other too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 16]
```diff
@@ -236,6 +236,149 @@ public boolean isAssignable(BType source, BType target) {
...
+                    for (BField field : ((BStructureType) target).getFields()) {
+                        if (field.getType().tag != mapConstraintTypeTag) {
+                            return false;
+                        }
+                    }
+                }
+                return true;
+            }
+        } else if (target.tag == TypeTags.JSON) {
+            if (((BJSONType) target).getConstraint().tag != TypeTags.NONE) {
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Shall we use fail-early patter.
i.e:
```ballerina
if (((BJSONType) target).getConstraint().tag == TypeTags.NONE) {
    return false;
}

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 17]
```diff
@@ -4127,6 +4168,110 @@ private static boolean checkIsType(BValue sourceVal, BType targetType) {
...
+            return checkIsLikeType(sourceValue, ((BJSONType) targetType).getConstrainedType());
+        } else if (sourceValue.getType().getTag() == TypeTags.ARRAY_TAG) {
+            BRefType<?>[] arrayValues = ((BRefValueArray) sourceValue).getValues();
+            for (int i = 0; i < ((BRefValueArray) sourceValue).size(); i++) {
+                if (!checkIsLikeType(arrayValues[i], targetType)) {
+                    return false;
+                }
+            }
+        }
+        return true;
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: should we check the value is a BMap value, and each value in the map is compatible with JSON?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java [hunk 18]
```diff
@@ -345,6 +348,33 @@ public BType getType() {
         return this.type;
     }
 
+    @Override
+    public void stamp(BType type) {
+        if (type.getTag() == TypeTags.JSON_TAG && ((BJSONType) type).getConstrainedType() != null) {
+            this.stamp(((BJSONType) type).getConstrainedType());
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BMap.java]: If constraint is null, each field should be stamped with JSON type.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 16 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 19]
```diff
@@ -4127,6 +4168,110 @@ private static boolean checkIsType(BValue sourceVal, BType targetType) {
...
+        for (Object object : ((BMap) sourceValue).getMap().entrySet()) {
+            Map.Entry valueEntry = (Map.Entry) object;
+            String fieldName = valueEntry.getKey().toString();
+
+            if (targetTypeField.containsKey(fieldName)) {
+                if (!checkIsLikeType(((BValue) valueEntry.getValue()), targetTypeField.get(fieldName))) {
+                    return false;
+                }
+            } else {
+                if (!((BRecordType) targetType).sealed) {
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: make sense. Fixed it.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 20]
```diff
@@ -4127,6 +4168,110 @@ private static boolean checkIsType(BValue sourceVal, BType targetType) {
...
+            for (int i = 0; i < ((BRefValueArray) sourceValue).size(); i++) {
+                if (!checkIsLikeType(arrayValues[i], targetType)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private static boolean checkIsLikeRecordType(BValue sourceValue, BType targetType) {
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: make sense. Fixed it.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java (hunk 17 above)
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: I don't see there is a need for this.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: required for optional field check..

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 21]
```diff
@@ -4127,6 +4168,110 @@ private static boolean checkIsType(BValue sourceVal, BType targetType) {
...
+            if (!(((BMap) sourceValue).getMap().containsKey(fieldName))) {
+                return false;
+            }
+        }
+
+        for (Object object : ((BMap) sourceValue).getMap().entrySet()) {
+            Map.Entry valueEntry = (Map.Entry) object;
+            String fieldName = valueEntry.getKey().toString();
+
+            if (targetTypeField.containsKey(fieldName)) {
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: required for optional field check..

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 22]
```diff
@@ -888,6 +893,42 @@ private static void handleFreezeBuiltinMethods(WorkerExecutionContext ctx, int o
...
+        int i = operands[0];
+        int j = operands[1];
+        int k = operands[2];
+
+        BRefType<?> valueToBeStamped = sf.refRegs[i];
+        BType stampType = ((TypeRefCPEntry) ctx.constPool[j]).getType();
+
+        if (stampType.getTag() == TypeTags.UNION_TAG) {
+            Predicate<BType> errorPredicate = e -> e.getTag() == TypeTags.ERROR_TAG;
+            ((BUnionType) stampType).getMemberTypes().removeIf(errorPredicate);
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Fixed.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 23]
```diff
@@ -888,6 +893,42 @@ private static void handleFreezeBuiltinMethods(WorkerExecutionContext ctx, int o
...
+        }
+
+        try {
+            valueToBeStamped.stamp(stampType);
+            sf.refRegs[k] = valueToBeStamped;
+        } catch (BallerinaException e) {
+            ctx.setError(BLangVMErrors.createError(ctx,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.INCOMPATIBLE_STAMP_OPERATION,
+                            valueToBeStamped.getType(), stampType)));
+            handleError(ctx);
```
[the remark this answers is not in the document]
Comment (mohanvive) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Fixed.

**Pull request merged**